File: exercises/capstone/ex02_gated_graph/ex02_gated_graph.cpp
// Exercise 02 -- gated boundaries in the graph (Step 2).
//
// Read exercises/capstone/ex02_gated_graph/README.md first.
#include "planning/airport/taxi_planner.hpp"

namespace planning::airport {

TaxiGraph buildGatedGraph(const TaxiGraph& raw, const ZoneLayer& layer) {
  // TODO(you): three passes.
  //
  // 1. Copy every raw vertex into the new graph, unchanged.
  //
  // 2. For each raw edge, walk it in steps of kZoneScanStep asking zoneAt()
  //    for the class.  Wherever the class changes, bisect to
  //    kZoneSplitTolerance and insert a new vertex there.  Give each resulting
  //    sub-edge the zone of its own midpoint, and carry over oneWay,
  //    maxWingspan and maxWeightTonnes from the parent.  The hotspot and
  //    closed flags are true if *any* sample of the sub-edge is inside one --
  //    a midpoint test misses a hotspot the edge only clips.
  //
  // 3. Call build(), then look at every vertex.  If its incident edges do not
  //    all agree on the zone class, it is a gate:
  //
  //      innerZone   the least restrictive class, by zonePriority()
  //      outerZone   the most restrictive
  //      gate        Hard when policyFor(outerZone).requiresClearance,
  //                  otherwise Soft
  //      protects    the idents of the zone polygon on the outer side
  //      holdShortId a holding position within half a metre, if there is one;
  //                  when there is, the vertex takes that line's name
  //
  // The directional part -- that a hard gate stops you going *in* and never
  // coming *out* -- is Exercise 05, and it is innerZone that makes it possible.
  (void)raw;
  (void)layer;
  return TaxiGraph{};
}

}  // namespace planning::airport

File: src/capstone/problems/ex02-gated-graph.ts
// Exercise 02 -- gated boundaries in the graph (Step 2).
// Brief: docs/capstone/ex02-gated-graph.md

import {
  distancePointSegment,
  policyFor,
  TaxiGraph,
  type ZoneLayer,
  zonePriority,
} from '../../airport/index.js';
import { ZONE_SCAN_STEP, ZONE_SPLIT_TOLERANCE } from '../types.js';
import { zoneAt } from './ex01-zone-layer.js';

export function buildGatedGraph(raw: TaxiGraph, layer: ZoneLayer): TaxiGraph {
  // TODO(you): the exercise that makes the invariant enforceable.  Three
  // passes over the raw chart:
  //
  //   1  Copy every vertex.  Then, for every raw edge, walk it in steps of
  //      ZONE_SCAN_STEP asking zoneAt() what zone the point is in.  Where the
  //      answer changes, bisect to ZONE_SPLIT_TOLERANCE and put a new vertex
  //      there.  Add one edge per piece, each tagged with the zone of its own
  //      midpoint and inheriting oneWay, maxWingspan and maxWeightTonnes.
  //
  //   2  hotspot and closed are properties of the *whole* sub-edge, not of its
  //      midpoint.  Taxiway F only clips the hotspot in its last fifty metres,
  //      and a midpoint test misses it.
  //
  //   3  After build(), a vertex whose incident edges do not all agree on the
  //      zone is a boundary, and a boundary on the graph is a gate.  Record
  //      innerZone (lowest zonePriority) and outerZone (highest); the gate is
  //      hard when policyFor(outerZone).requiresClearance, otherwise soft.
  //      Then find which runway is behind it, and whether a painted holding
  //      position sits on it -- that is what gives the gate a name a clearance
  //      can say out loud.
  void raw;
  void layer;
  void ZONE_SCAN_STEP;
  void ZONE_SPLIT_TOLERANCE;
  void zoneAt;
  void policyFor;
  void zonePriority;
  void distancePointSegment;
  return new TaxiGraph();
}
